bilge-impl/src/from_bits.rs

```diff
@@ -106,16 +106,43 @@ fn generate_enum(arb_int: TokenStream, enum_type: &Ident, match_arms: (Vec<Token
     }
 }
 
-fn generate_struct(arb_int: TokenStream, struct_type: &Ident) -> TokenStream {
+/// a single check per type is enough, so the checks can be deduped
+fn generate_assume_checks_for(ty: &Type, vec: &mut DedupedVec<TokenStream, String>) {
+    use Type::*;
+    match ty {
+        Path(_) => {
+            let assume = quote! { bilge::assume_filled::<#ty>(); };
+            vec.push(assume);
+        },
+        Tuple(tuple) => {
+            for elem in &tuple.elems {
+                generate_assume_checks_for(elem, vec)
+            }
+        },
+        Array(array) => generate_assume_checks_for(&array.elem, vec),
+        _ => unreachable(()),
+    }
+}
+
+fn generate_struct(arb_int: TokenStream, struct_type: &Ident, fields: &Fields) -> TokenStream {
     let const_ = if cfg!(feature = "nightly") {
         quote!(const)
     } else {
         quote!()
     };
 
+    let mut assumes = DedupedVec::new(TokenStream::to_string);
```

Comment:
what about itertools Dedup? now that we use that dep anyways

Reply:
that would be `UniqueBy` (`Dedup` needs the elements to be sorted).
and yeah we can do that now that I've had my daily fill of "let's implement this thing".